# 혈압 예측 모델 설계

## Load modules

In [1]:
import numpy as np

print("NumPy Version :{}".format(np.__version__))

NumPy Version :1.26.4


## Input and Label

In [2]:
# x_input은 입력 데이터입니다.
# 여기서는 사람의 나이 AGE를 의미합니다.
# 예: 25세, 35세, 45세, 55세, 65세, 73세
x_input = np.array(
    [25,25,25,35,35,35,45,45,45,55,55,55,65,65,65,73,73,73],
    dtype=np.float32
)

# labels는 정답 데이터입니다.
# 여기서는 각 나이에 해당하는 혈압 BP 값을 의미합니다.
# 머신러닝에서는 보통 정답값을 label 또는 y라고 부릅니다.
labels = np.array(
    [118,125,130,118,126,123,120,124,130,122,125,130,127,130,130,125.5,130,138],
    dtype=np.float32
)


# W는 기울기입니다.
# 선형회귀식 y = Wx + B 에서 W에 해당합니다.
# 나이가 1 증가할 때 혈압이 얼마나 증가하거나 감소하는지를 나타냅니다.
W = np.random.normal()

# B는 절편입니다.
# 선형회귀식 y = Wx + B 에서 B에 해당합니다.
# x가 0일 때의 기본값이라고 생각하면 됩니다.
B = np.random.normal()

## Hypothesis : Linear Equation
>### $h(x) = wx + b$

In [3]:
# Hypothesis는 가설 함수입니다.
# 쉽게 말하면 현재 W와 B를 이용해서 혈압을 예측하는 함수입니다.
# 선형회귀에서는 예측식을 보통 y = Wx + B 형태로 만듭니다.
def Hypothesis(x):
    return W * x + B

## Cost Function : Mean Squared Error (MSE)
>### cost(w, b) = $\frac{1}{m} \sum_{i=1}^{m}(h(x_{i})-y_{i})^{2}$

In [4]:
# Cost는 손실 함수입니다.
# 예측값과 실제 정답값이 얼마나 차이 나는지 계산합니다.
# 여기서는 MSE, 즉 평균제곱오차를 사용합니다.
# 오차를 제곱하는 이유는 음수 오차와 양수 오차가 서로 상쇄되지 않게 하기 위해서입니다.
def Cost():
    return np.mean((Hypothesis(x_input) - labels) ** 2)



## Gradient
>### $\frac{\partial}{\partial w}cost(w, b) = \frac{1}{m}  \sum_{i=1}^{m}(x_{i}(x_{i}w+(b-y_{i})))$
>### $\frac{\partial}{\partial b}cost(w, b) = \frac{1}{m}  \sum_{i=1}^{m}(x_{i}w - y_{i} + b)$



In [5]:
# Gradient는 기울기 값을 계산하는 함수입니다.
# 여기서 말하는 기울기는 그래프의 기울기가 아니라,
# W와 B를 어느 방향으로 얼마나 수정해야 손실이 줄어드는지를 알려주는 값입니다.
def Gradient(x, y):

    # grad_w는 W를 수정하기 위한 값입니다.
    # 예측값과 실제값의 차이가 W에 얼마나 영향을 주는지 계산합니다.
    grad_w = np.mean(x * (x * W + (B - y)))

    # grad_b는 B를 수정하기 위한 값입니다.
    # 예측값과 실제값의 차이가 B에 얼마나 영향을 주는지 계산합니다.
    grad_b = np.mean(W * x - y + B)

    # W용 기울기와 B용 기울기를 함께 반환합니다.
    return grad_w, grad_b


## Training
>### $\mu$ : Learning rate
>### $w = w - \mu\frac{\partial}{\partial w}cost(w, b)$
>### $b = b - \mu\frac{\partial}{\partial b}cost(w, b)$

In [9]:
# epochs는 전체 데이터를 몇 번 반복해서 학습할지 정하는 값입니다.
# 값이 클수록 더 많이 학습하지만 시간이 오래 걸립니다.
epochs = 200000


# learning_rate는 학습률입니다.
# W와 B를 한 번에 얼마나 크게 수정할지 정합니다.
# 너무 크면 학습이 불안정해지고, 너무 작으면 학습이 느려집니다.
learning_rate = 0.0005


# 본격적인 학습 과정입니다.
# cnt는 현재 몇 번째 학습 중인지 나타냅니다.
for cnt in range(0, epochs + 1):

    # 전체 학습 과정을 20등분해서 중간 결과를 출력합니다.
    # cost가 줄어드는지 확인하기 위한 코드입니다.
    if cnt % (epochs // 20) == 0:
        print(
            "[{:>5}] cost = {:>10.4}, W = {:>7.4}, B = {:>7.4}".format(
                cnt, Cost(), W, B
            )
        )

    # 현재 W와 B에서의 기울기를 계산합니다.
    grad_w, grad_b = Gradient(x_input, labels)

    # W를 손실이 줄어드는 방향으로 수정합니다.
    W -= learning_rate * grad_w

    # B를 손실이 줄어드는 방향으로 수정합니다.
    B -= learning_rate * grad_b

[    0] cost =      17.07, W =  0.1608, B =   118.2
[10000] cost =      17.07, W =  0.1608, B =   118.2
[20000] cost =      17.07, W =  0.1608, B =   118.2
[30000] cost =      17.07, W =  0.1608, B =   118.2
[40000] cost =      17.07, W =  0.1608, B =   118.2
[50000] cost =      17.07, W =  0.1608, B =   118.2
[60000] cost =      17.07, W =  0.1608, B =   118.2
[70000] cost =      17.07, W =  0.1608, B =   118.2
[80000] cost =      17.07, W =  0.1608, B =   118.2
[90000] cost =      17.07, W =  0.1608, B =   118.2
[100000] cost =      17.07, W =  0.1608, B =   118.2
[110000] cost =      17.07, W =  0.1608, B =   118.2
[120000] cost =      17.07, W =  0.1608, B =   118.2
[130000] cost =      17.07, W =  0.1608, B =   118.2
[140000] cost =      17.07, W =  0.1608, B =   118.2
[150000] cost =      17.07, W =  0.1608, B =   118.2
[160000] cost =      17.07, W =  0.1608, B =   118.2
[170000] cost =      17.07, W =  0.1608, B =   118.2
[180000] cost =      17.07, W =  0.1608, B =   118.2
[19

[200000]은 전체 학습을 200,000번 반복했다는 의미입니다. 즉, 데이터로 계속 수정하면서 최종 상태에 도달한 시점입니다.
cost = 17.07은 오차의 크기입니다. 이 값은 평균제곱오차(MSE)로, 모델이 예측한 혈압과 실제 혈압이 평균적으로 얼마나 차이 나는지를 나타냅니다. 값이 0에 가까울수록 완벽한 예측입니다. 현재 17.07이라는 값은 완전히 정확하지는 않지만, 데이터 특성상 어느 정도 오차가 있는 상태로 수렴했다는 의미입니다. 이 값을 직관적으로 보면, 제곱을 평균낸 값이므로 실제 평균 오차는 약 √17 ≈ 4 정도라고 이해할 수 있습니다. 즉, 평균적으로 약 ±4 mmHg 정도 틀린다고 해석할 수 있습니다.
W = 0.1609는 기울기입니다. 나이가 1 증가할 때 혈압이 약 0.1609만큼 증가한다는 의미입니다.
즉, 이 모델은 “나이가 많아질수록 혈압이 조금씩 증가한다”는 패턴을 학습한 것입니다.
B = 118.2는 절편입니다. 나이가 0일 때의 혈압 기준값으로, 실제 의미보다는 기준선 역할을 합니다.
이 값이 모델 전체를 위로 올리거나 내리는 역할을 합니다.
이 세 값을 종합하면, 최종 모델은 아래와 같은 식이 됩니다.
혈압 ≈ 0.1609 × 나이 + 118.2
이 식을 실제로 해석하면 다음과 같습니다.
나이가 증가할수록 혈압은 완만하게 증가하는 경향이 있으며, 기본 혈압 수준은 약 118 근처에서 시작한다는 의미입니다.
예를 들어 50세를 넣으면
혈압 ≈ 0.1609 × 50 + 118.2 ≈ 약 126 정도로 예측됩니다.
정리하면, 이 결과는 “모델이 데이터의 경향성은 잘 잡았지만, 개인별 변동 때문에 완벽하게 맞추지는 못한 상태”라고 해석하는 것이 가장 정확합니다.

## Predict

In [7]:
# Predict 함수는 새로운 나이를 넣었을 때 혈압을 예측하는 함수입니다.
def Predict(x):
    return Hypothesis(x)


# 50세의 혈압을 예측하기 위해 age 배열을 만듭니다.
age = np.array([50.0])

# 50세일 때 예측 혈압을 출력합니다.
print("{} Years : {:>7.4}mmHg".format(age[0], Predict(age)[0]))

50.0 Years :   126.2mmHg
